sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Product.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.util.TypeUtils
+import org.apache.spark.sql.types.{ AbstractDataType, DataType, DoubleType }
+
+
+/** Multiply numerical values within an aggregation group */
+case class Product(child: Expression, scale: Double = 1.0)
+    extends DeclarativeAggregate with ImplicitCastInputTypes {
+
+  override def children: Seq[Expression] = child :: Nil
+
+  override def nullable: Boolean = true
+
+  override def dataType: DataType = resultType
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(DoubleType)
+
+  override def checkInputDataTypes(): TypeCheckResult =
+    TypeUtils.checkForNumericExpr(child.dataType, "function product")
+
+  private val resultType = DoubleType
+
+  private lazy val product = AttributeReference("product", resultType)()
+
+  private lazy val one = Literal(1.0, resultType)
+
+  override lazy val aggBufferAttributes = product :: Nil
+
+  override lazy val initialValues: Seq[Expression] =
+    Seq(Literal(null, resultType))
+
+  override lazy val updateExpressions: Seq[Expression] = {
+    // Treat the result as null until we have seen at least one child value,
+    // whereupon the previous product is promoted to being unity.
+    // Each child value is implicitly multiplied by a scaling factor
+    // (1.0 by default) so that the client has some control over overflow
+    // when multiplying together many child values, without needing
+    // to explicitly rescale a Column beforehand.
+
+    val protoResult =
+      coalesce(product, one) * (scale match {
```

Comment:
Can you do:

```scala
coalesce(product, one) * scale match {
   ...
}
```

for simplicity?

BTW, this calculation is matched what @ueshin suggested offline. Can you double check @ueshin?

Reply:
I had tried that, but I don't think Scala's operator precedence allows this - the parentheses around the `match` clause seem necessary. The `protoResult` variable is necessary to cover `child.nullable` `if`-clause immediately after this.